runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -628,26 +804,45 @@ function LanguageTree:set_included_regions(new_regions)
         region[i] = { range:range(true) }
       end
     end
+    ---@cast region Range6[]
+
+    local i, exact = regions_inv_lookup(self._regions_inv, region)
+
+    if not exact then
+      if i then
+        self:_log(function()
+          return 'invalidating inexactly matched region', i, region_tostr(self._regions[i])
+        end)
+        regions_inv_remove(self._regions_inv, self._regions[i])
+      else
+        i = #self._regions + 1 -- this always gives an unoccupied index even if there are holes
+      end
+      self._regions[i] = region
+      regions_inv_insert(self._regions_inv, i, region)
+      self:_invalidate_region(i)
+    end
+    ---@cast i integer
+
+    touched[i] = true
   end
 
-  -- included_regions is not guaranteed to be list-like, but this is still sound, i.e. if
-  -- new_regions is different from included_regions, then outdated regions in included_regions are
-  -- invalidated. For example, if included_regions = new_regions ++ hole ++ outdated_regions, then
-  -- outdated_regions is invalidated by _iter_regions in else branch.
-  if #self:included_regions() ~= #new_regions then
-    -- TODO(lewis6991): inefficient; invalidate trees incrementally
-    for _, t in pairs(self._trees) do
-      self:_do_callback('changedtree', t:included_ranges(true), t)
+  -- Discard stale regions.
+  for i, _ in pairs(self._regions) do
+    if not touched[i] then
+      self:_discard_region(i)
     end
-    self._trees = {}
-    self:invalidate()
-  else
-    self:_iter_regions(function(i, region)
-      return vim.deep_equal(new_regions[i], region)
-    end)
   end
+end
 
-  self._regions = new_regions
+--- @param region Range6[]
+--- @return Range6[]
+local function prune_empty_ranges(region)
+  return vim
+    .iter(region)
+    :filter(function(r)
+      return r[3] ~= r[6]
+    end)
+    :totable()
```

Comment:
Hold that thought; `tbl_filter` is slated for deprecation in favor of -- wait for it -- `vim.iter():filter():totable()`.

Reply:
I disagree with that deprecation. We should not view `vim.iter` as a replacement for any `vim.tbl_*` functions. The entire point of `vim.iter` is for chaining operations.

Reply:
Then please make that case at https://github.com/neovim/neovim/issues/24572; I'm just the messenger.

Reply:
Looks like there is no _immediate_ plan: https://github.com/neovim/neovim/issues/24572#issuecomment-2074647896

Reply:
> The entire point of `vim.iter` is for chaining operations.

That rule came from nowhere. The point of vim.iter as described in https://github.com/neovim/neovim/issues/18585 is to provide an interface for iterables.

Reply:
That rule came from the author of `vim.iter`, which has been mentioned more than once in chat. Sorry to pull you in @gpanders but am I wrong here?

This is become a noticable problem since contributers have been reaching for `vim.iter` unnecessarily when almost all of the time a regular `pairs`/`ipairs` loop is enough.

Reply:
I agree that we do not need to use vim.iter every single time we need to perform a simple filter operation. Likewise for tbl_filter. I agree with @lewis6991, a regular for loop is often more readable and even faster.

vim.iter provides a generic iterator interface which is designed for working with pipelines. If your pipeline has only a single stage, then it’s just a complicated way of writing a for loop. Lua already has a “builtin” iterator concept so why not leverage that when it makes sense?

Reply:
Replaced with probably more efficient(?) in-place update
<https://github.com/neovim/neovim/compare/058d79cf7464732fc4fa5400994c614b49293c70..7b7736ced849114dcab5f3140e1565452d1ddb3d#diff-70b0f3fa7d6baf9e605400a9f0e10431183df821ec4096e8097eaa5f4d5531baR847-R858>

Reply:
> If your pipeline has only a single stage, then it’s just a complicated way of writing a for loop.

No disagreement there. I just want to make it clear that `vim.iter` is positioned for all kinds of iterables, to avoid the idea that we need table-only or list-only analogs for its functionality.